clouddriver-core/src/main/groovy/com/netflix/spinnaker/clouddriver/core/AccountHealthIndicator.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2020 Google, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.spinnaker.clouddriver.core;
+
+import com.google.common.collect.ImmutableMap;
+import com.netflix.spinnaker.clouddriver.security.AccountCredentials;
+import java.util.Optional;
+import javax.annotation.Nonnull;
+import org.springframework.boot.actuate.health.Health;
+import org.springframework.boot.actuate.health.HealthIndicator;
+import org.springframework.scheduling.annotation.Scheduled;
+
+/**
+ * This class represents a health indicator that reports on the status of individual cloud provider
+ * accounts.
+ *
+ * <p>It will always report a status of UP, to prevent issues with a single cloud provider from
+ * bringing down all of clouddriver, but any errors associated with individual accounts will appear
+ * in the detailed health information.
+ *
+ * @param <T> The type of account credentials this health indicator supports
+ */
+public abstract class AccountHealthIndicator<T extends AccountCredentials>
+    implements HealthIndicator {
+  @Nonnull private volatile Health health = new Health.Builder().up().build();
```

Comment:
Further inspiration to finish the concurrency chapter of EJ!

Reply:
Fair point, I think I may have been over-influenced by the fact that the class I was refactoring from used `AtomicReference` in a similar way.  Is there some guaranteed bound on when a write to a non-volatile variable would be visible by another thread?  I suppose it would never approach minutes (or even seconds) but just curious.

Reply:
No guaranteed bound that I know of, but yeah, I can't imagine it even being seconds...

Reply:
I'm definitely now musing about things I know very little about, but it [looks like the kernel context switches every 1.2–1.5µs](https://eli.thegreenplace.net/2018/measuring-context-switching-and-memory-overheads-for-linux-threads/)

Also from that link: "Every thread has some working set of memory, all or some of which is in the cache; when we switch to another thread, all this cache data becomes unneeded and is slowly flushed out, replaced by the new thread's data."

Of course, in a container where there are very few applications running, probably the cache flushes less often than in a VM, but still...

This could make an interesting java-users@ thread. Martin Buchholz, who held the title "JDK Concurrency Czar" at Sun, would almost certainly weigh in :)

Reply:
Thanks for the link!  Bookmarking it to read later when I have a block of time and won't have to context switch.